src/coreclr/jit/llvm.cpp

```diff
@@ -1175,38 +1201,76 @@ void Llvm::buildCall(GenTree* node)
-    {
-        // Cast pointer to int if necessary.  TODO-LLVM: candidate for lowering?
-        Value* sourceValue = castIfNecessary(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(cast->CastOp()->TypeGet()));
+    var_types castFromType  = genActualType(cast->CastOp());
+    var_types castToType    = cast->CastToType();
+    Value*    castFromValue = consumeValue(cast->CastOp(), getLlvmTypeForVarType(castFromType));
+    Value*    castValue     = nullptr;
 
-        mapGenTreeToValue(cast,
-            cast->IsUnsigned()
-            ? _builder.CreateZExt(sourceValue, getLlvmTypeForVarType(cast->CastToType()))
-            : _builder.CreateSExt(sourceValue, getLlvmTypeForVarType(cast->CastToType())));
-    }
-    else if (cast->TypeIs(TYP_INT, TYP_LONG) && cast->CastOp()->TypeIs(TYP_FLOAT, TYP_DOUBLE))
-    {
-        mapGenTreeToValue(cast,
-            cast->IsUnsigned()
-                ? _builder.CreateFPToUI(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(cast->CastToType()))
-                : _builder.CreateFPToSI(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(cast->CastToType())));
-    }
-    else
+    switch (castFromType)
     {
-        // TODO: other casts
-        failFunctionCompilation();
+        case TYP_BOOL:
+        case TYP_BYTE:
+        case TYP_UBYTE:
+        case TYP_SHORT:
+        case TYP_USHORT:
+        case TYP_INT:
+        case TYP_UINT:
+        case TYP_LONG:
+        case TYP_ULONG:
+            switch (castToType)
+            {
+                case TYP_BOOL:
+                {
+                    // nothing to do except map the source value to the destination GenTree
+                    castValue = _builder.CreateZExt(castFromValue, getLlvmTypeForVarType(TYP_INT));
+                    break;
+                }
+                case TYP_LONG:
+                {
+                    castValue = cast->IsUnsigned()
+                                    ? _builder.CreateZExt(castFromValue, getLlvmTypeForVarType(castToType))
+                                    : _builder.CreateSExt(castFromValue, getLlvmTypeForVarType(castToType));
+                    break;
+                }
+                case TYP_DOUBLE:
+                {
+                    castValue = cast->IsUnsigned()
+                                    ? _builder.CreateUIToFP(castFromValue, getLlvmTypeForVarType(castToType))
+                                    : _builder.CreateSIToFP(castFromValue, getLlvmTypeForVarType(castToType));
+                    break;
+                }
+                default:
+                    failFunctionCompilation(); // NYI
+            }
+            break;
+
+        case TYP_FLOAT:
+        case TYP_DOUBLE:
+            switch (castToType)
+            {
+                case TYP_DOUBLE:
+                {
+                    castValue = _builder.CreateFPCast(castFromValue, getLlvmTypeForVarType(TYP_DOUBLE));
+                    break;
+                }
```

Comment:
```suggestion
                case TYP_FLOAT:
                case TYP_DOUBLE:
                    castValue = _builder.CreateFPCast(castFromValue, getLlvmTypeForVarType(TYP_DOUBLE));
                    break;
```
(Assuming `case TYP_FLOAT` will work)